### PR TITLE
feat(sdk): Keep track of events read receipts in the timeline API

### PR DESCRIPTION
**Pull request by zecakeh**
This adds a list of read receipts for each event in the timeline and keeps it updated.

It also adds a map to allow to keep track of implicit read receipts (a read receipt because a member sent an event).

This is limited because we can't keep track of read receipts on events that don't create `TimelineEventItem`s. We would need to have an ordered list of all events in the timeline to have more reliable tracking.

Signed-off-by: Kévin Commaille <zecakeh@tedomum.fr>

### CODE REVIEW

File: crates/matrix-sdk/src/room/timeline/inner.rs [hunk 1]
```diff
@@ -577,17 +612,28 @@ impl ProfileProvider for room::Common {
             }
         }
     }
+
+    async fn read_receipts_for_event(&self, event_id: &EventId) -> IndexMap<OwnedUserId, Receipt> {
+        match self.event_receipts(ReceiptType::Read, ReceiptThread::Unthreaded, event_id).await {
+            Ok(receipts) => receipts.into_iter().collect(),
+            Err(e) => {
+                error!(%event_id, "Failed to get read receipts for event: {e}");
```
Comment (jplatte) [inline crates/matrix-sdk/src/room/timeline/inner.rs:620]: We're now using `Debug` formatting for tracing string fields everywhere.

```suggestion
                error!(?event_id, "Failed to get read receipts for event: {e}");
```

File: crates/matrix-sdk/src/room/timeline/mod.rs [hunk 2]
```diff
@@ -435,6 +436,18 @@ impl TimelineItem {
...
+    fn from(item: EventTimelineItem) -> Self {
+        Self::Event(item)
+    }
+}
+
+impl From<VirtualTimelineItem> for TimelineItem {
+    fn from(item: VirtualTimelineItem) -> Self {
+        Self::Virtual(item)
+    }
+}
```
Comment (jplatte) [inline crates/matrix-sdk/src/room/timeline/mod.rs:449]: Can this and its usages in existing code be a separate PR, to make this one a bit smaller?

File: crates/matrix-sdk/src/room/timeline/read_receipts.rs [hunk 3]
```diff
@@ -0,0 +1,228 @@
+// Copyright 2023 The Matrix.org Foundation C.I.C.
```
Comment (jplatte) [inline crates/matrix-sdk/src/room/timeline/read_receipts.rs:1]: I think you should put your own name here ^^

File: crates/matrix-sdk/src/room/timeline/read_receipts.rs [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+    receipt_type: ReceiptType,
+    receipt: &'a Receipt,
+}
+
+pub(super) fn handle_explicit_read_receipts(
+    receipt_event_content: ReceiptEventContent,
+    own_user_id: &UserId,
+    timeline_state: &mut TimelineInnerState,
+) {
+    for (event_id, receipt_types) in receipt_event_content.0 {
```
Comment (jplatte) [inline crates/matrix-sdk/src/room/timeline/read_receipts.rs:42]: Seems like we should implement `IntoIterator` for the event content types that are just newtypes over some collection.

File: crates/matrix-sdk/src/room/timeline/read_receipts.rs [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+        for (receipt_type, receipts) in receipt_types {
+            // We only care about read receipts here.
+            if !matches!(receipt_type, ReceiptType::Read | ReceiptType::ReadPrivate) {
+                continue;
+            }
+
+            for (user_id, receipt) in receipts {
+                let ReceiptThread::Unthreaded = receipt.thread else {
+                    continue;
+                };
```
Comment (jplatte) [inline crates/matrix-sdk/src/room/timeline/read_receipts.rs:52]:
```suggestion
                if receipt.thread != ReceiptThread::Unthreaded {
                    continue;
                }
```

File: crates/matrix-sdk/src/room/timeline/read_receipts.rs [hunk 6]
```diff
@@ -0,0 +1,228 @@
...
+                    receipt: &receipt,
+                };
+
+                if maybe_update_read_receipt(
+                    full_receipt,
+                    receipt_item_pos,
+                    is_own_user_id,
+                    &mut timeline_state.items,
+                    &mut timeline_state.users_read_receipts,
+                ) && !is_own_user_id
```
Comment (jplatte) [inline crates/matrix-sdk/src/room/timeline/read_receipts.rs:70]: Can you move this function call out into a binding?

File: crates/matrix-sdk/src/room/timeline/read_receipts.rs [hunk 7]
```diff
@@ -0,0 +1,228 @@
...
+                {
+                    // Update the new item pointed to by the user's read receipt.
+                    if let Some((pos, mut remote_event_item)) = receipt_item_pos.and_then(|pos| {
+                        timeline_state.items[pos]
+                            .as_event()
+                            .and_then(|e| e.as_remote())
+                            .map(|e| (pos, e.clone()))
```
Comment (jplatte) [inline crates/matrix-sdk/src/room/timeline/read_receipts.rs:77]:
```suggestion
                        let e = timeline_state.items[pos]
                            .as_event()?
                            .as_remote()?;
                        Some((pos, e.clone())
```

Comment (jplatte) [reply crates/matrix-sdk/src/room/timeline/read_receipts.rs:77]: Also moving the expression after the `=` of `if let` into a binding might make things more readable again.

File: crates/matrix-sdk/src/room/timeline/tests/invalid.rs [hunk 8]
```diff
@@ -88,6 +88,9 @@ async fn invalid_event_content() {
         }))
         .await;
 
+    let _read_receipt_update =
+        assert_matches!(stream.next().await, Some(VectorDiff::Set { value, .. }) => value);
+
```
Comment (jplatte) [inline crates/matrix-sdk/src/room/timeline/tests/invalid.rs:93]: I don't think it makes sense for all tests to have read receipt tracking on. Can you make it opt-in at the `TestTimeline` level (i.e. `TestTimeline::new().with_read_receipt_tracking()` or similar)?

File: crates/matrix-sdk/src/room/timeline/read_receipts.rs [hunk 9]
```diff
@@ -0,0 +1,228 @@
...
+    new_item_pos: Option<usize>,
+    is_own_user_id: bool,
+    timeline_items: &mut ObservableVector<Arc<TimelineItem>>,
+    users_read_receipts: &mut HashMap<OwnedUserId, HashMap<ReceiptType, (OwnedEventId, Receipt)>>,
+) -> bool {
+    let old_event_id = users_read_receipts
+        .get(receipt.user_id)
+        .and_then(|receipts| receipts.get(&receipt.receipt_type))
+        .map(|(event_id, _)| event_id);
+    if Some(receipt.event_id) == old_event_id.map(Deref::deref) {
```
[the remark this answers is not in the document]
Comment (zecakeh) [reply crates/matrix-sdk/src/room/timeline/read_receipts.rs:152]: It doesn't work, it's defined for `Option<T>`, not `Option<&T>`.

File: crates/matrix-sdk/src/room/timeline/mod.rs (hunk 2 above)
Comment (zecakeh) [reply crates/matrix-sdk/src/room/timeline/mod.rs:449]: Sure, I'll do it when this is approved and I can rebase everything, if that's okay.

File: crates/matrix-sdk/src/room/timeline/read_receipts.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (jplatte) [reply crates/matrix-sdk/src/room/timeline/read_receipts.rs:152]: Maybe `if old_event_id.map_or(false, |id| id == receipt.event_id)` will be clear with no need for `Deref` import? Not too hung up on changing this though.

File: crates/matrix-sdk/src/room/timeline/read_receipts.rs (hunk 3 above)
Comment (zecakeh) [reply crates/matrix-sdk/src/room/timeline/read_receipts.rs:1]: Is that mandatory? I don't really care about it. That's why I always put the Matrix foundation, it'll always be true at some point.

File: crates/matrix-sdk/src/room/timeline/mod.rs (hunk 2 above)
Comment (jplatte) [reply crates/matrix-sdk/src/room/timeline/mod.rs:449]: Oh to be clear, a separate PR that's merged first.

Comment (zecakeh) [reply crates/matrix-sdk/src/room/timeline/mod.rs:449]: Oh it's my fault, I read "PR" and thought "commit".

File: crates/matrix-sdk/src/room/timeline/read_receipts.rs [hunk 10]
```diff
@@ -47,7 +47,7 @@ pub(super) fn handle_explicit_read_receipts(
             }
 
             for (user_id, receipt) in receipts {
-                let ReceiptThread::Unthreaded = receipt.thread else {
+                if receipt.thread != ReceiptThread::Unthreaded {
                     continue;
                 };
```
Comment (jplatte) [inline crates/matrix-sdk/src/room/timeline/read_receipts.rs:52]:
```suggestion
                }
```

File: crates/matrix-sdk/src/room/timeline/read_receipts.rs (hunk 3 above)
Comment (jplatte) [reply crates/matrix-sdk/src/room/timeline/read_receipts.rs:1]: Hm, I guess probably not? Have asked.

Comment (jplatte) [reply crates/matrix-sdk/src/room/timeline/read_receipts.rs:1]: It seems like we really want you to put in your own name 😄

File: crates/matrix-sdk/src/room/timeline/read_receipts.rs [hunk 11]
```diff
@@ -0,0 +1,228 @@
...
+        receipt: &receipt,
+    };
+
+    if maybe_update_read_receipt(
+        new_receipt,
+        Some(item_pos),
+        is_own_event,
+        timeline_items,
+        users_read_receipts,
+    ) && !is_own_event
```
Comment (jplatte) [inline crates/matrix-sdk/src/room/timeline/read_receipts.rs:123]:
```suggestion
    let receipt_saved = maybe_update_read_receipt(
        new_receipt,
        Some(item_pos),
        is_own_event,
        timeline_items,
        users_read_receipts,
    );
    if receipt_saved && !is_own_event
```

File: crates/matrix-sdk/src/room/timeline/tests/mod.rs [hunk 12]
```diff
@@ -245,6 +264,24 @@ impl TestTimeline {
         })
     }
 
+    fn make_receipt_event_content(
+        &self,
+        receipts: impl IntoIterator<Item = (OwnedEventId, ReceiptType, OwnedUserId, ReceiptThread)>,
+    ) -> ReceiptEventContent {
+        let mut ev_content = ReceiptEventContent(BTreeMap::new());
+        for (event_id, receipt_type, user_id, thread) in receipts.into_iter() {
```
Comment (jplatte) [inline crates/matrix-sdk/src/room/timeline/tests/mod.rs:272]:
```suggestion
        for (event_id, receipt_type, user_id, thread) in receipts {
```

**Review by jplatte — APPROVED** [review]
Small style suggestions but I think we can merge after those are handled, and git conflicts are resolved.

**Comment by poljar** [comment]
> Small style suggestions but I think we can merge after those are handled, and git conflicts are resolved.

Can we merge this now or is there something else we need to do?
